### PR TITLE
Implemented API evolution pattern

### PR DESCRIPTION
**Pull request by marioradu05**

### CODE REVIEW

File: src/App/src/Message.php [hunk 1]
```diff
@@ -44,4 +44,5 @@ class Message
     public const VALIDATOR_REQUIRED_FIELD         = 'This field is required and cannot be empty.';
     public const VALIDATOR_REQUIRED_FIELD_BY_NAME = '%s is required and cannot be empty.';
     public const VALIDATOR_REQUIRED_UPLOAD        = 'A file must be uploaded first.';
+    public const RESTRICTION_DEPRECATION          = 'Cannot use both `%s` and `%s` attributes on the same object.';
```
Comment (alexmerlin) [inline src/App/src/Message.php]: We try keep messages ordered alphabetically, so let's move this line up, after the one with `RESOURCE_NOT_ALLOWED`.
Also, I see that I misplaced `MAIL_NOT_SENT_TO` - let's move that too after `INVALID_VALUE`.

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        RequestHandlerInterface $handler
+    ): ResponseInterface {
+        $response    = $handler->handle($request);
+        $routeResult = $request->getAttribute(RouteResult::class);
+        if (! $routeResult instanceof RouteResult || $routeResult->isFailure()) {
+            return $response;
+        }
+
+        $reflectionHandler = null;
+        $routeMiddleware   = $routeResult->getMatchedRoute()->getMiddleware();
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        $routeMiddleware   = null;
        $matchedRoute      = $routeResult->getMatchedRoute();
        if ($matchedRoute !== false) {
            $routeMiddleware = $matchedRoute->getMiddleware();
        }
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        if ($routeMiddleware instanceof LazyLoadingMiddleware) {
+            $reflectionMiddlewareClass = new ReflectionClass($routeMiddleware->middlewareName);
+            if ($reflectionMiddlewareClass->implementsInterface(RequestHandlerInterface::class)) {
+                $reflectionHandler = $reflectionMiddlewareClass;
+            }
+        } else {
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        } elseif ($routeMiddleware !== null) {
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                    Message::RESTRICTION_DEPRECATION,
+                    self::RESOURCE_DEPRECATION_ATTRIBUTE,
+                    self::METHOD_DEPRECATION_ATTRIBUTE
+                )
+            );
+        }
+
+        $sunset = '';
+        $link   = '';
+        if ($attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE] ?? '') {
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        if (array_key_exists(self::RESOURCE_DEPRECATION_ATTRIBUTE, $attributes)) {
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            $sunset = $attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE]['sunset'];
+            $link   = $attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE]['link'];
+        }
+
+        if ($attributes[self::METHOD_DEPRECATION_ATTRIBUTE] ?? '') {
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        if (array_key_exists(self::METHOD_DEPRECATION_ATTRIBUTE, $attributes)) {
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            throw new DeprecationConflictException(
+                sprintf(
+                    Message::RESTRICTION_DEPRECATION,
+                    self::RESOURCE_DEPRECATION_ATTRIBUTE,
+                    self::METHOD_DEPRECATION_ATTRIBUTE
+                )
+            );
+        }
+
+        $sunset = '';
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        $sunset = null;
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        $link   = '';
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        $link   = null;
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            $sunset = $attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE]['sunset'];
+            $link   = $attributes[self::RESOURCE_DEPRECATION_ATTRIBUTE]['link'];
+        }
+
+        if ($attributes[self::METHOD_DEPRECATION_ATTRIBUTE] ?? '') {
+            $sunset = $attributes[self::METHOD_DEPRECATION_ATTRIBUTE]['sunset'];
+            $link   = $attributes[self::METHOD_DEPRECATION_ATTRIBUTE]['link'];
+        }
+
+        if ($sunset) {
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        if ($sunset !== null) {
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            $response = $response->withHeader('sunset', $sunset);
+        }
+
+        if ($link) {
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
        if ($link !== null) {
```

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            $response = $response->withHeader('link', $link);
+        }
+
+        return $response;
+    }
+
+    public function getAttributes(ReflectionClass|ReflectionMethod $reflection, string $type): array
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php:118]:
```suggestion
    /**
     * @param class-string $type
     */
    public function getAttributes(ReflectionClass|ReflectionMethod $reflection, string $type): array
```

---

To fix this Psalm error:

> ArgumentTypeCoercion: Argument 1 of ReflectionClass::getAttributes expects class-string|null, but parent type string provided

File: src/App/src/Middleware/DeprecationMiddleware.php [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        $response    = $handler->handle($request);
+        $routeResult = $request->getAttribute(RouteResult::class);
+        if (! $routeResult instanceof RouteResult || $routeResult->isFailure()) {
+            return $response;
+        }
+
+        $reflectionHandler = null;
+        $routeMiddleware   = $routeResult->getMatchedRoute()->getMiddleware();
+        if ($routeMiddleware instanceof LazyLoadingMiddleware) {
+            $reflectionMiddlewareClass = new ReflectionClass($routeMiddleware->middlewareName);
```
Comment (alexmerlin) [inline src/App/src/Middleware/DeprecationMiddleware.php]:
```suggestion
            /** @var class-string $routeMiddleware */
            $routeMiddleware           = $routeMiddleware->middlewareName;
            $reflectionMiddlewareClass = new ReflectionClass($routeMiddleware);
```

File: test/Unit/App/Attribute/MethodDeprecationTest.php [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+{
+    public function testInvalidDateThrowsException(): void
+    {
+        $class = new class {
+            #[MethodDeprecation(
+                sunset: 'invalid-01-01',
+                link: 'test-link',
+                deprecationReason: 'test-deprecation-reason',
+            )]
+            public function test()
```
Comment (alexmerlin) [inline test/Unit/App/Attribute/MethodDeprecationTest.php]:
```suggestion
            public function test(): void
```

File: test/Unit/App/Attribute/MethodDeprecationTest.php [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+    public function testValidDatePassesValidation(): void
+    {
+        $class = new class {
+            #[MethodDeprecation(
+                sunset: '2038-01-01',
+                link: 'test-link',
+                deprecationReason: 'test-deprecation-reason',
+            )]
+            public function test()
```
Comment (alexmerlin) [inline test/Unit/App/Attribute/MethodDeprecationTest.php]:
```suggestion
            public function test(): void
```

File: test/Unit/App/Middleware/DeprecationMiddlewareTest.php [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        $this->request  = $this->createMock(ServerRequestInterface::class);
+        $this->response = new EmptyResponse();
+        $this->subject  = new Subject();
+    }
+
+    /**
+     * @throws ReflectionException
+     * @throws Exception
+     */
+    public function testThrowsDeprecationConflictException()
```
Comment (alexmerlin) [inline test/Unit/App/Middleware/DeprecationMiddlewareTest.php]:
```suggestion
    public function testThrowsDeprecationConflictException(): void
```

**Review by alexmerlin — CHANGES_REQUESTED** [review]

File: src/App/src/Middleware/DeprecationMiddleware.php (hunk 3 above)
Comment (marioradu05) [reply src/App/src/Middleware/DeprecationMiddleware.php]: `$routeMiddleware` can't be null so I checked if it's an instance of `MiddlewarePipe `which implements `MiddlewareInterface`.

**Review by alexmerlin — APPROVED** [review]

**Review by arhimede — APPROVED** [review]

**Pull request merged**